distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Add_Running.enso

```diff
@@ -0,0 +1,44 @@
+from Standard.Base import all
+import Standard.Base.Errors.Common.Unsupported_Argument_Types
+import Standard.Base.Errors.Illegal_Argument.Illegal_Argument
+
+import project.Column.Column
+import project.Set_Mode.Set_Mode
+import project.Sort_Column.Sort_Column
+import project.Table.Table
+import project.Internal.Add_Row_Number
+import project.Internal.Java_Problems
+import project.Internal.Problem_Builder.Problem_Builder
+import project.Internal.Table_Helpers
+from project.Errors import Duplicate_Output_Column_Names
+import project.Value_Type.Value_Type
+
+polyglot java import java.lang.ArithmeticException
+polyglot java import org.enso.table.data.column.storage.numeric.LongRangeStorage
+polyglot java import org.enso.table.operations.AddRunning
+
+## PRIVATE
+add_running : Statistic -> (Text | Integer) -> Text -> Vector (Text | Integer | Regex) | Text | Integer | Regex -> Vector (Text | Sort_Column) | Text -> Problem_Behavior -> Table
+add_running table (statistic:Statistic=Statistic.Count) (of:Text|Integer=0) (as:Text='') (group_by:(Vector | Text | Integer | Regex)=[]) (order_by:(Vector | Text)=[]) (on_problems:Problem_Behavior=Problem_Behavior.Report_Warning) =
+    check_running_support [statistic] <|
+        of_col = table.at of
+        new_name = if as.is_empty then 'Running ' + statistic.to_text + ' of ' + of_col.name else as   
+        case statistic of
+            Statistic.Count ->
+                Add_Row_Number.add_row_number table new_name 1 1 group_by order_by on_problems
```

Comment:
Would it make sense to actually move the `Add_Row_Number` logic into `AddRunning`?

Then we could have `add_row_number` be just a shorthand for `add_running Count` and keep all of this logic in one place.

It feels to me like currently we are duplicating the logic for iterating over the rows grouped by a key and potentially ordered - once in `Add_Row_Number` and once in `AddRunning`. It feels like ideally this should be done in one abstraction, unless there is some reason that would make it unnecessarily complicated that I'm missing?

Reply:
Ah so I realize that currently one obstacle is that `AddRunning` deals with `DoubleStorage` but `Add_Row_Number` is based on `LongStorage`, as pointed out in one of the earlier comments.

That indeed is blocking us from re-using this single abstraction. So I guess it's totally OK to keep the 2 for this PR.

But ideally, once we get to the value type of the running calculations depend on the inputs, I think it should be possible to merge the `Add_Row_Number` logic into the `AddRunning` and re-use the `RunningIterator` concept for it. I think that would be good because the relatively complex ordering/grouping logic would then be just in one place, instead of two like right now.

Reply:
And I realized one more obstacle that is probably worth considering - if `add_row_number` has _no ordering nor grouping_, it completes in O(1) because it creates a 'virtual' column, that also uses O(1) memory.

If we decide to merge the logic as I suggested, we probably would like to preserve this nice feature, which may still require a special case (but I guess that could be just an `if` inside of `add_row_number` itself for a special codepath).

Reply:
Yes once I have resolved the return type for min/max etc. I hope I can fold the add_row_number logic back in with this. But future MR.